File: src/net/t3kt/tctrl/schema/ParentSchemaNode.java
package net.t3kt.tctrl.schema;

import com.google.common.collect.ImmutableList;
import com.google.protobuf.Message;

public abstract class ParentSchemaNode<S extends Message> extends SchemaNode<S> {
    protected ParentSchemaNode(S spec) {
        super(spec);
    }

    public abstract SchemaGroupList getChildGroups();

    public abstract ImmutableList<ModuleSchema> getChildModules();
}

File: win/TctrlCore/Schema/ParentSchemaNode.cs
using System.Collections.Generic;
using System.Linq;
using Google.Protobuf;
using Tctrl.Schema;

namespace Tctrl.Core.Schema {
    public abstract class ParentSchemaNode<S> : SchemaNode<S> where S : IMessage {
        protected ParentSchemaNode(S spec, IEnumerable<ModuleSpec> childModules) : base(spec) {
            ChildModules = childModules.Select(moduleSpec => new ModuleSchema(moduleSpec)).ToList().AsReadOnly();
        }
        public IReadOnlyList<ModuleSchema> ChildModules { get; }
    }
}